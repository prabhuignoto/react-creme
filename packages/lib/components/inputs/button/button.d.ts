import * as React from 'react';
import { ButtonProps } from './button-model';
declare const Button: React.ForwardRefExoticComponent<
  ButtonProps & React.RefAttributes<HTMLDivElement>
>;
export { Button };
