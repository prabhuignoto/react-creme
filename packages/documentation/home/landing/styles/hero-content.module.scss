@use '@design/theme.scss';

.content {
  border-radius: 8px;
  padding: 3rem;
}

@mixin grid {
  column-gap: 2rem;
  display: grid;
  grid-template-columns: 1fr 2fr;
  justify-items: center;
}

@media (min-width: 768px) {
  .content {
    max-width: 600px;
  }
}

@media (min-width: 1024px) {
  .content {
    display: flex;
    flex-direction: column;
    max-width: 800px;
  }
}

@media (min-width: 1280px) {
  .content {
    @include grid;
    max-width: 1000px;
  }
}

@media (min-width: 1440px) {
  .content {
    @include grid;
    max-width: 1240px;
  }
}

@media (min-width: 1920px) {
  .content {
    @include grid;
    max-width: 1700px;
  }
}

@media (min-width: 2560px) {
  .content {
    @include grid;
    max-width: 2000px;
  }
}

.left_column {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  min-width: 400px;
}

.right_column {
  align-items: center;
  align-self: flex-end;
  display: grid;
  gap: 1rem;
  grid-template-columns: 1fr 1fr;
  justify-items: center;
  width: 100%;
}

.header {
  color: theme.$primary;
  font-size: 3rem;
  font-weight: 500;
  margin: 0.2rem 0;
  text-transform: uppercase;
}

.sub_text {
  font-size: 2rem;
  font-weight: 300;
  margin: 1rem 0;

  &.dark {
    color: #fff;
  }
}

.get_started_btn {
  background-color: var(--rc-secondary-color-hex);
  border: 0;
  border-radius: 4px;
  color: var(--rc-color-white);
  font-size: 1.2rem;
  margin-top: 3rem;
  padding: 1rem;
  width: 200px;
}

.buttons {
  display: flex;
  margin: 3rem 0 0;

  > :first-child {
    margin-right: 1rem;
  }
}
