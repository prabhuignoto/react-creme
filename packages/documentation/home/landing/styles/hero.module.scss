.hero {
  align-items: flex-start;
  display: flex;
  justify-content: center;
  width: 100%;

  @media (min-width: 1rem) {
    min-height: 1200px;
  }

  @media (min-width: 43rem) {
    min-height: 980px;
  }

  @media (min-width: 768px) {
    min-height: 1080px;
  }

  @media (min-width: 1280px) {
    min-height: 820px;
  }

  &:not(.dark) {
    background: linear-gradient(
      to left bottom,
      rgba(#bfdbf7, 1) 20%,
      rgba(#e1e5f2, 1)
    );
  }

  &.dark {
    background: linear-gradient(
      to right bottom,
      rgba(#b8c0ff, 0.3) 10%,
      rgba(#1e1e1e, 0.9)
    );
  }
}
