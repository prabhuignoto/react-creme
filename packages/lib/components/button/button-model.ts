import React, { CSSProperties, KeyboardEvent, MouseEvent } from 'react';

export interface ButtonProps {
  accent?: 'flat' | 'rounded';
  border?: boolean;
  children?: React.ReactNode;
  disabled?: boolean;
  focusable?: boolean;
  label?: string;
  onClick?: (ev?: MouseEvent | KeyboardEvent) => void;
  primary?: boolean;
  size?: 'sm' | 'md' | 'lg';
  style?: CSSProperties;
  type?: 'primary' | 'default' | 'danger' | 'icon' | 'progress';
}
