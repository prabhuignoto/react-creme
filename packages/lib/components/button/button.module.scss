@use '@design/core.scss';
@use '@design/theme.scss';
@use '@design/button.scss';
@use '@design/animate.scss';
@use 'sass:map';

$types: (primary, default, danger);
$sizes: (sm, md, lg);

.btn {
  &.no_border {
    border: 0;
    box-shadow: none;
  }

  &:not(.icon) {
    min-width: 50px;
  }

  &:not(.flat):not(.dark) {
    @extend %border;
  }

  &.dark:not(.flat) {
    @extend %border-radius;
  }

  &:not(:active) {
    transition: transform 80ms ease-in;
  }

  &:active:not(.disabled) {
    transform: scale(0.95);
  }
}

@each $value in $types {
  .#{$value} {
    @extend %button-#{$value};

    &.disabled {
      @extend %disabled;
      border: 1px solid transparent;
    }
  }
}

@each $size in $sizes {
  .#{$size} {
    @extend %text-#{$size};
    @extend %button-#{$size};

    .icon_container {
      @extend %icon-#{$size};
    }

    .label {
      @extend %font-#{$size};
    }
  }
}

.label {
  @extend %text-no-wrap;
  text-transform: capitalize;
}

@each $size in $sizes {
  .#{$size} {
    .label {
      @extend %font-#{$size};
    }

    .icon {
      @extend %icon-#{$size};
    }
  }
}

.icon {
  border-radius: 50%;

  &.disabled {
    @extend %disabled;
  }

  &:not(.disabled):not(.dark) {
    background: theme.$light-bg;
  }

  &:hover:not(.disabled):not(.dark) {
    background: theme.$primary;
    color: theme.$white;
  }

  &.dark:not(.disabled) {
    background: theme.$dark-control-bg;
  }

  &:hover.dark:not(.disabled) {
    background: theme.$primary;
  }

  &:hover.dark:not(.disabled) svg {
    color: theme.$white;
  }
}

.icon_container + .label {
  margin-left: 0.25rem;
}

.progress_wrapper {
  margin-right: 0.5rem;
}
