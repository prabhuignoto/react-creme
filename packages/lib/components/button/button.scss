@use '../../design/layout.scss';
@use '../../design/theme.scss';
@use '../../design/button.scss';
@use '../../design/animate.scss';
@use 'sass:map';

$types: (primary, default, danger);
$sizes: (sm, md, lg);

.rc-btn {
  &.rc-btn-no-border {
    border: 0;
    box-shadow: none;
  }

  &:not(.rc-btn-icon) {
    min-width: 50px;
  }
}

@each $value in $types {
  .rc-btn-#{$value} {
    @extend %button-#{$value};

    &.rc-disabled {
      @extend %disabled;
      border: 1px solid theme.$chinese-gray;
    }
  }
}

@each $size in $sizes {
  .rc-btn-#{$size} {
    @extend %text-#{$size};
    @extend %button-#{$size};

    .rc-btn-icon-container {
      @extend %icon-#{$size};
    }

    .rc-btn-label {
      @extend %font-#{$size};
    }
  }
}

.rc-btn-label {
  @extend %text-no-wrap;
  text-transform: capitalize;
}

@each $size in $sizes {
  .rc-btn-#{$size} {
    .rc-btn-label {
      @extend %font-#{$size};
    }

    .rc-btn-icon {
      @extend %icon-#{$size};
    }
  }
}

.rc-btn-icon {
  border-radius: 50%;
  padding: 0.25rem;

  &.rc-disabled {
    @extend %disabled;
  }

  &:not(.rc-disabled) {
    background: theme.$light-bg;
  }

  &:hover:not(.rc-disabled) {
    background: theme.$secondary;
  }
}

.rc-btn-icon-container + .rc-btn-label {
  margin-left: 0.25rem;
}

.rc-btn-progress-wrapper {
  margin-right: 0.5rem;
}
