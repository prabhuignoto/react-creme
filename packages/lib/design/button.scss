@use 'theme.scss';
@use 'core.scss';

%button {
  @extend %center;
  background: none;
  cursor: pointer;
  transition: background-color 50ms ease-in;
  white-space: nowrap;

  &:not(.dark) {
    @extend %border;
  }

  &.dark {
    @extend %border-radius;
  }

  svg {
    color: theme.$primary;
  }

  &:hover svg {
    color: theme.$white;
  }
}

@mixin get-button-type($bg, $bg-hover, $color, $color-hover) {
  background: $bg;
  color: $color;

  svg {
    color: $color;
  }

  &:hover {
    svg {
      color: $color-hover;
    }
  }

  &:hover:not(.rc-disabled) {
    background: $bg-hover;
    box-shadow: none;
    color: $color-hover;
  }
}

%button-primary {
  @include get-button-type(
    rgba(theme.$primary-rgb, 0.85),
    theme.$primary,
    theme.$white,
    theme.$white
  );
}

%button-danger {
  @include get-button-type(
    rgba(theme.$crimson, 0.9),
    theme.$crimson,
    theme.$white,
    theme.$white
  );
  border: 1px solid theme.$crimson;
}

%button-default {
  &.dark {
    @include get-button-type(
      theme.$cod-gray,
      theme.$primary,
      theme.$white,
      theme.$white
    );
  }

  &:not(.dark) {
    @extend %shadow-inset-small;
    @include get-button-type(
      theme.$white,
      theme.$primary,
      theme.$primary,
      theme.$white
    );
  }
}

%button-icon {
  @extend %button;
  border: 0;
}

%button-icon:hover {
  svg {
    color: theme.$white;
  }
}

@mixin get-button($size, $border) {
  @extend %button;

  &:not(.icon) {
    @if $size == small {
      height: 20px;
      padding: calc(20px * 0.25) calc(20px * 0.3);
    } @else if $size == large {
      height: 30px;
      padding: calc(30px * 0.25) calc(30px * 0.3);
    } @else if $size == medium {
      height: 25px;
      padding: calc(25px * 0.25) calc(25px * 0.3);
    }
  }

  &.icon {
    border: 0;
  }
}

%button-sm {
  @include get-button(small, true);
}

%button-md {
  @include get-button(medium, true);
}

%button-lg {
  @include get-button(large, true);
}

%button-sm-no-border {
  @include get-button(small, false);
}
%button-lg-no-border {
  @include get-button(large, false);
}
