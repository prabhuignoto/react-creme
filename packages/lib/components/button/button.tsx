import classNames from "classnames";
import React, { useMemo, useRef } from "react";
import { useFocus } from "../common/effects/useFocus";
import { CircularProgress } from "../progress/circular-progress";
import { ButtonModel } from "./button-model";
import "./button.scss";

const Button: React.FunctionComponent<ButtonModel> = ({
  border = true,
  children,
  disabled = false,
  focusable = true,
  type = "default",
  label = "",
  onClick,
  size = "sm",
  style = {},
}) => {
  // classes
  const buttonClass = useMemo(
    () =>
      classNames(
        {
          "rc-btn-default": type === "progress",
          "rc-btn-no-border": !border,
          "rc-disabled": disabled,
        },
        [`rc-btn-${size}`, `rc-btn-${type}`, "rc-btn"]
      ),
    [disabled]
  );

  // setup for focus

  const ref = useRef(null);

  if (focusable) {
    useFocus(ref, onClick);
  }

  const focusableProps = useMemo(
    () => ({
      tabIndex: focusable ? 0 : -1,
    }),
    []
  );

  // handler for button click
  const handleClick = () => !disabled && onClick?.();

  return (
    <button
      className={buttonClass}
      onClick={handleClick}
      ref={ref}
      role="button"
      style={style}
      {...focusableProps}
      aria-label={label}
    >
      {type === "progress" && !disabled && (
        <span className="rc-btn-progress-wrapper">
          <CircularProgress size={"xs"} />
        </span>
      )}
      {children && <span className="rc-btn-icon-container">{children}</span>}
      {label && type !== "icon" && (
        <span className="rc-btn-label">{label}</span>
      )}
    </button>
  );
};

export { Button };
