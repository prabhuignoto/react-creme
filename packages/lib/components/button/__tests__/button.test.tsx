import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import { vi } from 'vitest';
import { Button } from '../button';
import styles from '../button.module.scss';

const handler = vi.fn();

describe('Button', () => {
  it('should render default', () => {
    const { container } = render(<Button />);

    const button = container.firstChild;

    expect(button).toBeInTheDocument();
    expect(button).toHaveClass(styles.default);
  });

  it('should render label', () => {
    const { getByText } = render(<Button label="My Button" />);
    expect(getByText('My Button')).toBeInTheDocument();
  });

  it('should render size', () => {
    const { getByRole } = render(<Button label="My Button" size="lg" />);
    expect(getByRole('button')).toBeInTheDocument();
    expect(getByRole('button')).toHaveClass(styles.lg);
  });

  it('should render button snapshot', () => {
    const { getByRole } = render(<Button label="My Button" />);
    expect(getByRole('button')).toMatchSnapshot();
  });

  it('should render disabled button', () => {
    const handler = vi.fn();

    const { getByRole } = render(
      <Button label="My Button" disabled onClick={handler} />
    );

    expect(getByRole('button')).toHaveClass(styles.disabled);

    fireEvent.click(getByRole('button'));

    expect(handler).not.toBeCalled();
  });

  it('should call handler', () => {
    const { getByText } = render(
      <Button label="My Button" onClick={handler} />
    );

    fireEvent.click(getByText('My Button'));

    expect(handler).toBeCalled();
  });

  it('should call handler via keyboard action', async () => {
    const { getByText } = render(
      <Button label="My Button" onClick={handler} />
    );

    fireEvent.keyDown(getByText('My Button'), { key: 'Enter' });

    await waitFor(() => {
      expect(handler).toBeCalled();
    });
  });

  it('should have focus', () => {
    const { getByRole } = render(
      <Button label="My Button" onClick={handler} focusable />
    );

    fireEvent.focus(getByRole('button'));

    expect(getByRole('button')).toHaveAttribute('tabIndex', '0');
  });
});
