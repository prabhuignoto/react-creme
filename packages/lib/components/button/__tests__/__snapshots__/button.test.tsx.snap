// Vitest Snapshot v1

exports[`Button > should render button snapshot 1`] = `
<div
  aria-label="My Button"
  class="_btn_sm_13o6m_1 _btn_default_13o6m_9 _btn_13o6m_1"
  role="button"
  style="outline: none; position: relative;"
  tabindex="0"
>
  <span
    class="_btn_label_13o6m_13"
  >
    My Button
  </span>
  <span
    class="rc-focus-ring"
    style="width: 6px; height: 6px;"
  />
</div>
`;
